File: src/main/java/com/kingmang/lazurite/parser/AST/Statements/ForStatement.java
package com.kingmang.lazurite.parser.AST.Statements;


import com.kingmang.lazurite.parser.AST.Expressions.Expression;
import com.kingmang.lazurite.parser.AST.InterruptableNode;
import com.kingmang.lazurite.patterns.visitor.ResultVisitor;
import com.kingmang.lazurite.patterns.visitor.Visitor;
import lombok.AllArgsConstructor;

@AllArgsConstructor
public final class ForStatement extends InterruptableNode implements Statement {
    
    public final Statement initialization;
    public final Expression termination;
    public final Statement increment;
    public final Statement statement;

    @Override
    public void execute() {
        super.interruptionCheck();
        for (initialization.execute(); termination.eval().asInt() != 0; increment.execute()) {
            try {
                statement.execute();
            } catch (BreakStatement bs) {
                break;
            } catch (ContinueStatement cs) {
                // continue;
            }
        }
    }
    
    @Override
    public void accept(Visitor visitor) {
        visitor.visit(this);
    }

    @Override
    public <R, T> R accept(ResultVisitor<R, T> visitor, T t) {
        return visitor.visit(this, t);
    }

    @Override
    public String toString() {
        return "for " + initialization + ", " + termination + ", " + increment + " " + statement;
    }
}

File: src/main/java/com/kingmang/lazurite/parser/ast/statements/ForStatement.kt
package com.kingmang.lazurite.parser.ast.statements

import com.kingmang.lazurite.parser.ast.InterruptableNode
import com.kingmang.lazurite.parser.ast.expressions.Expression
import com.kingmang.lazurite.patterns.visitor.ResultVisitor
import com.kingmang.lazurite.patterns.visitor.Visitor

data class ForStatement(
    val initialization: Statement,
    val termination: Expression,
    val increment: Statement,
    val statement: Statement
) : InterruptableNode(), Statement {
    override fun execute() {
        super.interruptionCheck()
        this.initialization.execute()
        while (this.termination.eval().asInt() != 0) {
            try {
                this.statement.execute()
            } catch (bs: BreakStatement) {
                break
            } catch (cs: ContinueStatement) {
                // continue;
            }
            this.increment.execute()
        }
    }

    override fun accept(visitor: Visitor) =
        visitor.visit(this)

    override fun <R, T> accept(visitor: ResultVisitor<R, T>, input: T): R? =
        visitor.visit(this, input)

    override fun toString(): String =
        "for ${this.initialization}, ${this.termination}, ${this.increment} ${this.statement}"
}
